pkg/server/plugin/keymanager/base/keymanagerbase.go

```diff
@@ -0,0 +1,253 @@
+type Base struct {
+	impl Impl
+
+	mu      sync.RWMutex
+	entries map[string]*KeyEntry
+}
+
+func New(impl Impl) *Base {
+	return &Base{
+		impl:    impl,
+		entries: make(map[string]*KeyEntry),
+	}
+}
+
+func (m *Base) SetEntries(entries []*KeyEntry) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.entries = entriesMapFromSlice(entries)
+}
+
+func (m *Base) GenerateKey(ctx context.Context, req *keymanager.GenerateKeyRequest) (*keymanager.GenerateKeyResponse, error) {
+	if req.KeyId == "" {
+		return nil, m.newError("key id is required")
+	}
+	if req.KeyAlgorithm == keymanager.KeyAlgorithm_UNSPECIFIED_KEY_ALGORITHM {
+		return nil, m.newError("key algorithm is required")
+	}
+
+	newEntry, err := m.generateKeyEntry(req.KeyId, req.KeyAlgorithm)
+	if err != nil {
+		return nil, err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	oldEntry, hasEntry := m.entries[req.KeyId]
+
+	m.entries[req.KeyId] = newEntry
+
+	if m.impl.WriteFn != nil {
+		if err := m.impl.WriteFn(ctx, entriesSliceFromMap(m.entries)); err != nil {
+			delete(m.entries, req.KeyId)
+			if hasEntry {
+				m.entries[req.KeyId] = oldEntry
+			}
+			return nil, err
+		}
+	}
+
+	return &keymanager.GenerateKeyResponse{
+		PublicKey: clonePublicKey(newEntry.PublicKey),
+	}, nil
+}
+
+func (m *Base) GetPublicKey(ctx context.Context, req *keymanager.GetPublicKeyRequest) (*keymanager.GetPublicKeyResponse, error) {
+	if req.KeyId == "" {
+		return nil, m.newError("key id is required")
+	}
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	resp := new(keymanager.GetPublicKeyResponse)
+	entry := m.entries[req.KeyId]
+	if entry != nil {
+		resp.PublicKey = clonePublicKey(entry.PublicKey)
+	}
+
+	return resp, nil
+}
+
+func (m *Base) GetPublicKeys(ctx context.Context, req *keymanager.GetPublicKeysRequest) (*keymanager.GetPublicKeysResponse, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	resp := new(keymanager.GetPublicKeysResponse)
+	for _, entry := range entriesSliceFromMap(m.entries) {
+		resp.PublicKeys = append(resp.PublicKeys, clonePublicKey(entry.PublicKey))
+	}
+
+	return resp, nil
+}
+
+func (m *Base) SignData(ctx context.Context, req *keymanager.SignDataRequest) (*keymanager.SignDataResponse, error) {
+	if req.KeyId == "" {
+		return nil, m.newError("key id is required")
+	}
+	if req.HashAlgorithm == keymanager.HashAlgorithm_UNSPECIFIED_HASH_ALGORITHM {
+		return nil, m.newError("hash algorithm is required")
+	}
+
+	privateKey := m.getPrivateKey(req.KeyId)
+	if privateKey == nil {
+		return nil, m.newError("no such key %q", req.KeyId)
+	}
+
+	signer, ok := privateKey.(crypto.Signer)
+	if !ok {
+		return nil, m.newError("keypair %q not usable for signing", req.KeyId)
+	}
+
+	signature, err := signer.Sign(rand.Reader, req.Data, crypto.Hash(req.HashAlgorithm))
+	if err != nil {
+		return nil, m.newError("keypair %q signing operation failed: %v", req.KeyId, err)
+	}
+
+	return &keymanager.SignDataResponse{
+		Signature: signature,
+	}, nil
+}
+
+func (m *Base) getPrivateKey(id string) crypto.PrivateKey {
```

Comment:
I'm curious about the intention behind having functionality like this be wrapped up in a base package instead of being written directly into the keymanager plugins, or implemented instead as helpers? Presumably HSM etc won't be able to furnish the private key... is this base package applicable to all imaginable types of keymanager implementations'?

Reply:
so, originally I had a single disk plugin, where the path was optional, and if not provided wouldn't do persistence. but some comments on previous PR's (i.e. where I added optional persistence to the `memory` ServerCA`) made me think that it was desirable to have separate `memory` and `disk` keymanager plugins. The implementation is about 99% similar. This path lead to the smallest boilerplate between the two implementations.

Reply:
Then intention was only for this to be shared between the memory and disk plugins.